### PR TITLE
Fix Visual Studio bit-field alignment

### PR DESCRIPTION
**Pull request by kroening**

### CODE REVIEW

File: src/ansi-c/c_typecheck_type.cpp [hunk 1]
```diff
@@ -927,7 +927,13 @@ void c_typecheck_baset::typecheck_compound_body(
   if(type.id()==ID_struct)
     add_padding(to_struct_type(type), *this);
   else if(type.id()==ID_union)
+  {
+    // Visual Studio ignores padding on unions
+    if(config.ansi_c.mode==configt::ansi_ct::flavourt::VISUAL_STUDIO)
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_type.cpp]: Nit pick: space around `==`

**Review by tautschnig — DISMISSED** [review]

**Comment by smowton** [comment]
Should update the description-- the test added doesn't relate to packed unions

File: src/ansi-c/padding.cpp [hunk 2]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+  unsignedbv_typet padding_type(pad_bits);
+
+  struct_typet::componentt component;
+  component.type() = padding_type;
+  component.set_name("$pad" + std::to_string(where - components.begin()));
+  component.set_is_padding(true);
+
+  auto it = components.insert(where, component);
+  it++;
+  return it;
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: `return std::next(components.insert(where, component));` (though `return ++components.insert(where, component);` would do as well)

File: src/ansi-c/padding.cpp [hunk 3]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+struct_typet::componentst::iterator pad_bit_field(
+  struct_typet::componentst &components,
+  struct_typet::componentst::iterator where,
+  std::size_t pad_bits)
+{
+  c_bit_field_typet padding_type(unsignedbv_typet(pad_bits), pad_bits);
+
+  struct_typet::componentt component;
+  component.type() = padding_type;
+  component.set_name(
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: Use `component(padding_type, "$bit_field_pad" + ...)` for proper RAII.

File: src/ansi-c/padding.cpp [hunk 4]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
   return result;
 }
 
-void add_padding(struct_typet &type, const namespacet &ns)
+optionalt<std::size_t>
+underlying_width(const c_bit_field_typet &type, const namespacet &ns)
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: Should be `static`.

File: src/ansi-c/padding.cpp [hunk 5]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+    if(c_enum_type.id() == ID_c_enum)
+      return c_enum_type.subtype().get_int(ID_width);
+    else
+      return {};
+  }
+  else
+    return {};
+}
+
+struct_typet::componentst::iterator pad_bit_field(
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: Should be `static`

File: src/ansi-c/padding.cpp [hunk 6]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+  component.set_name(
+    "$bit_field_pad" + std::to_string(where - components.begin()));
+  component.set_is_padding(true);
+
+  auto it = components.insert(where, component);
+  it++;
+  return it;
+}
+
+struct_typet::componentst::iterator pad(
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: Should be `static`

File: src/ansi-c/padding.cpp [hunk 7]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+  struct_typet::componentst &components,
+  struct_typet::componentst::iterator where,
+  std::size_t pad_bits)
+{
+  unsignedbv_typet padding_type(pad_bits);
+
+  struct_typet::componentt component;
+  component.type() = padding_type;
+  component.set_name("$pad" + std::to_string(where - components.begin()));
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: See above for proper RAII.

File: src/ansi-c/padding.cpp [hunk 8]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+  it++;
+  return it;
+}
+
+struct_typet::componentst::iterator pad(
+  struct_typet::componentst &components,
+  struct_typet::componentst::iterator where,
+  std::size_t pad_bits)
+{
+  unsignedbv_typet padding_type(pad_bits);
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: `const`

File: src/ansi-c/padding.cpp [hunk 9]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+  else
+    return {};
+}
+
+struct_typet::componentst::iterator pad_bit_field(
+  struct_typet::componentst &components,
+  struct_typet::componentst::iterator where,
+  std::size_t pad_bits)
+{
+  c_bit_field_typet padding_type(unsignedbv_typet(pad_bits), pad_bits);
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: `const`

File: src/ansi-c/padding.cpp [hunk 10]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+
+  struct_typet::componentt component;
+  component.type() = padding_type;
+  component.set_name(
+    "$bit_field_pad" + std::to_string(where - components.begin()));
+  component.set_is_padding(true);
+
+  auto it = components.insert(where, component);
+  it++;
+  return it;
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: See below for `std::next`

File: src/ansi-c/padding.cpp [hunk 11]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+  component.type() = padding_type;
+  component.set_name("$pad" + std::to_string(where - components.begin()));
+  component.set_is_padding(true);
+
+  auto it = components.insert(where, component);
+  it++;
+  return it;
+}
+
+void add_padding_msvc(struct_typet &type, const namespacet &ns)
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: `static`

File: src/ansi-c/padding.cpp [hunk 12]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
 {
   struct_typet::componentst &components=type.components();
 
-  // First do padding for bit-fields to make them
-  // appear on byte boundaries.
+  std::size_t bit_field_bits = 0, underlying_bits = 0;
+  mp_integer offset = 0;
+
+  bool is_packed=type.get_bool(ID_C_packed);
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: `const`, spacing around `=`

File: src/ansi-c/padding.cpp [hunk 13]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+    // if we continue a bit-field with size>0 and the same underlying width
+
+    if(
+      it->type().id() == ID_c_bit_field &&
+      to_c_bit_field_type(it->type()).get_width() != 0 &&
+      underlying_width(to_c_bit_field_type(it->type()), ns).value_or(0) ==
+        underlying_bits)
+    {
+      // do not add padding, but count the bits
+      std::size_t width = to_c_bit_field_type(it->type()).get_width();
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: `const`, though I would really eliminate that temporary

File: src/ansi-c/padding.cpp [hunk 14]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+      bit_field_bits += width;
+    }
+    else
+    {
+      // pad up any remaining bit field
+      if(underlying_bits != 0 && (bit_field_bits % underlying_bits) != 0)
+      {
+        std::size_t pad_bits =
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: `const`

File: src/ansi-c/padding.cpp [hunk 15]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+          underlying_bits - (bit_field_bits % underlying_bits);
+        it = pad_bit_field(components, it, pad_bits);
+        offset += (bit_field_bits + pad_bits) / 8;
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: Use `config.ansi_c.char_width` instead of `8`

File: src/ansi-c/padding.cpp [hunk 16]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+      {
+        const mp_integer a = alignment(it->type(), ns);
+        if(a > 1)
+        {
+          const mp_integer displacement = offset % a;
+
+          if(displacement != 0)
+          {
+            mp_integer pad_bytes = a - displacement;
+            std::size_t pad_bits = integer2size_t(pad_bytes * 8);
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: `config.ansi_c.char_width` instead of `8`

File: src/ansi-c/padding.cpp [hunk 17]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+      if(!is_packed)
+      {
+        const mp_integer a = alignment(it->type(), ns);
+        if(a > 1)
+        {
+          const mp_integer displacement = offset % a;
+
+          if(displacement != 0)
+          {
+            mp_integer pad_bytes = a - displacement;
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: `const`

File: src/ansi-c/padding.cpp [hunk 18]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+          }
+        }
+      }
+
+      // do we start a new bit field?
+      if(it->type().id() == ID_c_bit_field)
+      {
+        underlying_bits =
+          underlying_width(to_c_bit_field_type(it->type()), ns).value_or(0);
+        std::size_t width = to_c_bit_field_type(it->type()).get_width();
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: `const`, though as above: I think that temporary can be eliminated

File: src/ansi-c/padding.cpp [hunk 19]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+    }
+  }
+
+  // Add padding at the end?
+  // Bit-field
+  if(underlying_bits != 0 && (bit_field_bits % underlying_bits) != 0)
+  {
+    std::size_t pad = underlying_bits - (bit_field_bits % underlying_bits);
+    pad_bit_field(components, components.end(), pad);
+    offset += (bit_field_bits + pad) / 8;
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: `config.ansi_c.char_width` instead of `8`

File: src/ansi-c/padding.cpp [hunk 20]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+          offset += size;
+      }
+    }
+  }
+
+  // Add padding at the end?
+  // Bit-field
+  if(underlying_bits != 0 && (bit_field_bits % underlying_bits) != 0)
+  {
+    std::size_t pad = underlying_bits - (bit_field_bits % underlying_bits);
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: `const`

File: src/ansi-c/padding.cpp [hunk 21]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+
+  // alignment of the struct
+  // Note that this is done even if the struct is packed.
+  const mp_integer a = alignment(type, ns);
+  const mp_integer displacement = offset % a;
+
+  if(displacement != 0)
+  {
+    mp_integer pad_bytes = a - displacement;
+    std::size_t pad_bits = integer2size_t(pad_bytes * 8);
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: `config.ansi_c.char_width` instead of `8`

File: src/ansi-c/padding.cpp [hunk 22]
```diff
@@ -100,15 +100,176 @@ mp_integer alignment(const typet &type, const namespacet &ns)
...
+  }
+
+  // alignment of the struct
+  // Note that this is done even if the struct is packed.
+  const mp_integer a = alignment(type, ns);
+  const mp_integer displacement = offset % a;
+
+  if(displacement != 0)
+  {
+    mp_integer pad_bytes = a - displacement;
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: `const`

File: src/ansi-c/padding.cpp [hunk 23]
```diff
@@ -226,20 +368,10 @@ void add_padding(struct_typet &type, const namespacet &ns)
...
-        struct_typet::componentt component;
-        component.type()=padding_type;
-        component.set_name("$pad"+std::to_string(padding_counter++));
-        component.set_is_padding(true);
-
-        it=components.insert(it, component);
-        it++; // skip over
-
-        offset+=pad;
+        mp_integer pad_bytes = a - displacement;
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: `const`

File: src/ansi-c/padding.cpp [hunk 24]
```diff
@@ -226,20 +368,10 @@ void add_padding(struct_typet &type, const namespacet &ns)
...
+        std::size_t pad_bits = integer2size_t(pad_bytes * 8);
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: `const`, `config.ansi_c.char_width` instead of `8`

File: src/ansi-c/padding.cpp [hunk 25]
```diff
@@ -277,22 +409,22 @@ void add_padding(struct_typet &type, const namespacet &ns)
...
-      unsignedbv_typet padding_type(width);
-
-      // we insert after any final 'flexible member'
-      struct_typet::componentt component;
-      component.type()=padding_type;
-      component.set_name("$pad"+std::to_string(padding_counter++));
-      component.set_is_padding(true);
-
-      components.push_back(component);
+      mp_integer pad_bytes = max_alignment - displacement;
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp:416]: `const`

File: src/ansi-c/padding.cpp [hunk 26]
```diff
@@ -277,22 +409,22 @@ void add_padding(struct_typet &type, const namespacet &ns)
...
+      std::size_t pad_bits = integer2size_t(pad_bytes * 8);
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp]: `const`, `config.ansi_c.char_width` instead of `8`

File: src/ansi-c/padding.cpp [hunk 27]
```diff
@@ -277,22 +409,22 @@ void add_padding(struct_typet &type, const namespacet &ns)
...
+      pad(components, components.end(), pad_bits);
     }
   }
 }
 
+void add_padding(struct_typet &type, const namespacet &ns)
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp:424]: `static`

File: src/ansi-c/padding.cpp [hunk 28]
```diff
@@ -313,6 +445,19 @@ void add_padding(union_typet &type, const namespacet &ns)
...
   }
 
+  if(config.ansi_c.mode == configt::ansi_ct::flavourt::VISUAL_STUDIO)
+  {
+    // Visual Studio pads up to the underlying width of
+    // any bit field.
+    for(const auto &c : type.components())
+      if(c.type().id() == ID_c_bit_field)
+      {
+        auto w = underlying_width(to_c_bit_field_type(c.type()), ns);
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp:461]: `const`

File: src/ansi-c/padding.cpp [hunk 29]
```diff
@@ -313,6 +445,19 @@ void add_padding(union_typet &type, const namespacet &ns)
...
+        if(w.has_value() && w.value() > max_alignment_bits)
```
Comment (tautschnig) [inline src/ansi-c/padding.cpp:462]: Could use `w.value_or(0)`

**Review by tautschnig — COMMENTED** [review]
Lots of nit-picking...

**Comment by kroening** [comment]
All done

**Review by tautschnig — APPROVED** [review]
A couple of comments are still open, but assuming these are addressed this looks good to me.

File: src/ansi-c/padding.cpp (hunk 27 above)
Comment (kroening) [reply src/ansi-c/padding.cpp:424]: That one is called from the outside.

Comment (tautschnig) [reply src/ansi-c/padding.cpp:424]: Ah, ok, I should have checked.

**Review by peterschrammel — APPROVED** [review]
LGTM

**Pull request merged**